judge/models/problem.py

```diff
@@ -255,6 +255,43 @@ def is_accessible_by(self, user, skip_contest_problem_check=False):
     def is_subs_manageable_by(self, user):
         return user.is_staff and user.has_perm('judge.rejudge_submission') and self.is_editable_by(user)
 
+    @classmethod
+    def get_visible_problems(cls, user):
```

Comment:
should be merged with `problems_list`

Reply:
Most of this code can be kept, but `is_restricted` and `see_restricted_problem` has to be merged in.